warehouse/query-core/src/main/java/datawave/query/transformer/ExcerptTransform.java

```diff
@@ -0,0 +1,197 @@
+     * @param document
+     *            the document
+     * @return the phrase indexes
+     */
+    private PhraseIndexes getPhraseIndexes(Document document) {
+        Content content = (Content) document.get(PHRASE_INDEXES_ATTRIBUTE);
+        return PhraseIndexes.from(content.getContent());
+    }
+    
+    /**
+     * Add the excerpts to the document as part of {@value #HIT_EXCERPTS}.
+     * 
+     * @param excerpts
+     *            the excerpts to add
+     * @param document
+     *            the document
+     */
+    private void addExcerptsToDocument(Set<String> excerpts, Document document) {
+        Attributes attributes = new Attributes(true);
+        for (String excerpt : excerpts) {
+            Content content = new Content(excerpt, document.getMetadata(), true);
+            attributes.add(content);
+        }
+        document.put(HIT_EXCERPTS, attributes);
+    }
+    
+    /**
+     * Get the excerpts.
+     * 
+     * @param phraseIndexes
+     *            the pre-identified phrase offsets
+     * @return the excerpts
+     */
+    private Set<String> getExcerpts(PhraseIndexes phraseIndexes, Document document) {
+        phraseIndexes = getOffsetPhraseIndexes(phraseIndexes);
+        if (phraseIndexes.isEmpty()) {
+            return Collections.emptySet();
+        }
+        
+        // Construct the required range for this document.
+        Key metadata = document.getMetadata();
+        Key startKey = new Key(metadata.getRow(), new Text("datatype\0uid"));
+        Key endKey = startKey.followingKey(PartialKey.ROW_COLFAM);
+        Range range = new Range(startKey, true, endKey, false);
+        
+        // Fetch the excerpts.
+        Set<String> excerpts = new HashSet<>();
+        for (String field : phraseIndexes.getFields()) {
+            Collection<Pair<Integer,Integer>> indexes = phraseIndexes.getIndices(field);
+            for (Pair<Integer,Integer> indexPair : indexes) {
+                String excerpt = getExcerpt(field, indexPair.getValue0(), indexPair.getValue1(), range);
+                excerpts.add(excerpt);
+            }
+        }
+        return excerpts;
+    }
+    
+    private String getExcerpt(String field, int start, int end, Range range) {
+        Map<String,String> options = new HashMap<>();
+        options.put(TermFrequencyExcerptIterator.FIELD_NAME, field);
+        options.put(TermFrequencyExcerptIterator.START_OFFSET, String.valueOf(start));
+        options.put(TermFrequencyExcerptIterator.END_OFFSET, String.valueOf(end));
+        TermFrequencyExcerptIterator iterator = new TermFrequencyExcerptIterator();
+        try {
+            iterator.init(source, options, env);
+            iterator.seek(range, Collections.emptyList(), false);
+            if (iterator.hasTop()) {
+                iterator.getTopKey();
+            }
+            // TODO - seek to phrase
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to scan for excerpts", e);
```

Comment:
@ivakegg do you want an exception thrown here, or swallowed and logged instead? If an exception, is there a type you prefer?

Reply:
This is probably fine as is.  I do not want it swallowed.

Reply:
Okay.